packages/gatsby/src/commands/build.js

```diff
@@ -144,7 +169,13 @@ module.exports = async function build(program: BuildArgs) {
   // we need to save it again to make sure our latest state has been saved
   await db.saveState()
 
-  const pagePaths = [...store.getState().pages.keys()]
+  const pagePaths = process.env.GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES
```

Comment:
I think it would be more readable if we compare webpackCompilationHash here and not in `getChangedPageDataKeys`:

The actual condition that dictates when we rebuild only part of changes is then more obvious, as well as it is confusing to why we are passing compliation hash to `getChangedPageDataKeys` function (which name doesn't indicate to have anything to do with webpack compilation hash.

I would also add a note like this before assigning `pagePaths` (might need to be written in nicer manner):
```
// Rebuild subset of pages if user opt into it and if there were no source files (for example components) changes since last build. Otherwise rebuild all pages
```